File: subscriber.py
"""Definition of the Subscriber class."""
# Author(s): Pierre Abraham Mulamba
# Date of creation (modification): 20200224 (20200224)
# Usage: import the subscriber
#        create an object sub = Subscriber("102030", "Pierre", "Abraham", 20)

from dataclasses import dataclass


# from borrow import Borrow


@dataclass(init=True, repr=True)
class Subscriber(object):
    """
    Definition of a Subscriber with the following attributes:
    id_number: str
    first_name: str
    last_name: str
    age: int
    borrowers: list[Borrow]
    """
    __slots__ = ['id_number', 'first_name', 'last_name', 'age']
    id_number: str
    first_name: str
    last_name: str
    age: int

    # borrowers: list

    # Accessors methods
    @property
    def get_id_number(self) -> str:
        """
        Access and returns the user id_number
        :return: id_number
        """
        return self.id_number

    @property
    def get_first_name(self) -> str:
        """
        Access and returns the attribute first_name of the class Subscriber
        :return: the attribute first_name
        """
        return self.first_name

    @property
    def get_last_name(self) -> str:
        """
        Accesses and returns the attribute last_name of the class Subscriber
        :return: the attribute last_name
        """
        return self.last_name

    @property
    def get_age(self) -> int:
        """
        Accesses and returns the attribute age of the class Subscriber
        :return: age
        """
        return self.age

    # @property
    # def get_borrowers(self) -> Borrow:
    #    """
    #    Returns the list of borrowers
    #    :return:
    #    """
    #    try:
    #        return self.borrowers
    #    except Exception as exception_list_borrowers:
    #        print("{}".format(exception_list_borrowers))
    #    finally:
    #        pass

    # @property
    # def get_number_of_borrowed_items_by_subsciber(self) -> int:
    #    """
    #    Returns the number of borrowed items by the subscriber
    #    :return:
    #    """
    #    try:
    #        return self.borrowers.count(self)
    #    except

    # Mutators methods
    def set_id_number(self, id_number):
        """
        Mutator method to set the attribute id_number of the class Subscriber
        :param id_number:str
        """
        self.id_number = id_number

    def set_first_name(self, first_name):
        """
        Mutator method to set the attribute first_name of the class Subscriber
        :param first_name: str
        :return: void
        """
        self.first_name = first_name

    def set_last_name(self, last_name):
        """
        Mutator method to set the attribute last_name of the class Attribute
        :param last_name: str
        :return: void
        """
        self.last_name = last_name

    def set_age(self, age):
        """
        Set the age
        :param age: int
        :return: void
        """
        self.age = age

    def info(self):
        """
        Method to access and return the attributes of the class Subscriber
        :return: void
        """
        return f"{self.get_first_name}, {self.get_last_name}. {self.get_age} y.o. #{self.get_id_number}"

File: Subscriber.java
/**
 * File: Subscriber.java
 * Author(s): Pierre Abraham Mulamba
 * Date of creation (modification): 20200627 (20200627)
 * Description: creating the class Subscriber with default values using
 * the builder design pattern method
 * link: https://stackoverflow.com/questions/222214/
 * managing-constructors-with-many-parameters-in-java/222295#222295
 * Compilation: javac *.java
 * Run: java Main
 * Note: jdk (Java development Kits)
 *       jre (Java Runtime Environment)
 *       se (Standard Edition)
 * 
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.smart.tech;

/**
 * Attributes
 * ==========
 * idNumber
 * firstName
 * lastName
 * age
 * count
 * 
 * Methods
 * =========
 * Subscriber(): ctor
 * Subscriber(idNumber, firstName, lastName, age)
 * Accessor 
 * Mutator
 * print
 * @author pamulamba
 */
public class Subscriber extends Object {
    private String idNumber;
    private String firstName;
    private String lastName;
    private int age;
    private static int count;
    
    // Ctor
    public Subscriber(){}
    
    public Subscriber(String id_number, String first_name, String last_name, int age_){
        idNumber = id_number;
        firstName = first_name;
        lastName = last_name;
        age = age_;
        ++count;
    }
    
    // Accessor methods
    public String getIdNumber(){
        return idNumber;
    }
    
    public String getFirstName(){
        return firstName;
    }
    
    public String getLastName(){
        return lastName;
    }
    
    public int getAge(){
        return age;
    }
    
    public static int getCount(){
        return count;
    }
    
    // Copy ctor
    public Subscriber(Subscriber subscriber){
        this.idNumber = subscriber.getIdNumber();
        this.firstName = subscriber.getFirstName();
        this.lastName = subscriber.getLastName();
        this.age = subscriber.getAge();
    }
    
    // Mutator methods
    public void setIdNumber(String id_number){
        idNumber = ((0 < id_number.length() && id_number.length() <= 6) ? id_number : "" );
    }
    
    public void setFirstName(String first_name){
        firstName = ((0 < first_name.length() && first_name.length() <= 10 ) ? first_name : "");
    }
    
    public void setLastName(String last_name){
        lastName = ((0 < last_name.length() && last_name.length() <= 10 ) ? last_name : "");
    }
    
    public void setAge(int age_){
        age = ((3 <= age_ && age_ <= 120) ? age_ : 0);
    }
    
    // Builder design pattern methods
    public Subscriber idNumber(String id_number){
        this.idNumber = id_number;
        return this;
    }
    
    public Subscriber firstName(String first_name){
        this.firstName = first_name;
        return this;
    }
    
    public Subscriber lastName(String last_name){
        this.lastName = last_name;
        return this;
    }
    
    public Subscriber age(int age_){
        this.age = age_;
        return this;
    }
    
    // Printing methods
     public void print(){
         System.out.printf("\nSubscriber: #%s. %s , %s. %d\n",
                 getIdNumber(), getLastName(), getFirstName(), getAge());
    }
   
} // End class Subscriber
